### ISSUE TITLE
nnx.scan working

### ISSUE THREAD

**Issue by ak24watch** [google/flax#4780]
I want to know how nnx.scan is different from jax.lax.scan. with examples.

**Comment by vfdev-5** [comment]
@ak24watch you can check this guide: https://flax.readthedocs.io/en/latest/guides/linen_to_nnx.html#transformations

**Comment by ak24watch** [comment]
I am having doubt that in nnx.scan what is in_axis and out_axis ans what is nnx.carry, can you please explain with simple examples. 


1) is in_axis first argument is carry and second one **xs** can mention which axis to iterate over.

2) out_axis is  forst argument is carry and second argument is for stacking on which axis for  **y**.

Is Understanding correct, please correct me  with examples if wrong.

**Comment by ak24watch** [comment]
> @ak24watch you can check this guide: https://flax.readthedocs.io/en/latest/guides/linen_to_nnx.html#transformations

I am having doubt that in nnx.scan what is in_axis and out_axis ans what is nnx.carry, can you please explain with simple examples. 


1) is in_axis first argument is carry and second one **xs** can mention which axis to iterate over.

2) out_axis is  forst argument is carry and second argument is for stacking on which axis for  **y**.

Is Understanding correct, please correct me  with examples if wrong.

**Comment by mohsinm-dev** [comment]
@ak24watch Your understanding is **correct**! Let me confirm with examples.

  ### Key Differences
  - **`jax.lax.scan`**: Works with arrays only, manual state management
  - **`nnx.scan`**: Works with NNX Modules, automatic state management

  ### Simple Example
  ```python
  import jax.numpy as jnp
  from flax import nnx

  # Running sum example
  @nnx.scan(in_axes=(nnx.Carry, 0), out_axes=(nnx.Carry, 0))
  def running_sum(carry, x):
      new_carry = carry + x
      output = new_carry
      return new_carry, output

  numbers = jnp.array([1.0, 2.0, 3.0, 4.0])
  final_sum, all_sums = running_sum(0.0, numbers)

  print(final_sum)    # 10.0
  print(all_sums)     # [1.0, 3.0, 6.0, 10.0]

  Real Power: Neural Network Layers

  class Block(nnx.Module):
      def __init__(self, dim, rngs):
          self.linear = nnx.Linear(dim, dim, rngs=rngs)

      def __call__(self, x):
          return nnx.relu(self.linear(x))

  class ScanMLP(nnx.Module):
      def __init__(self, dim, n_layers, rngs):
          # Create vectorized layers
          @nnx.split_rngs(splits=n_layers)
          @nnx.vmap(axis_size=n_layers)
          def create_block(rngs):
              return Block(dim, rngs=rngs)

          self.layers = create_block(rngs)

      def __call__(self, x):
          # Scan through layers
          @nnx.scan(in_axes=(nnx.Carry, 0), out_axes=(nnx.Carry, 0))
          def apply_layer(x, layer):
              return layer(x), layer(x)

          final_x, _ = apply_layer(x, self.layers)
          return final_x

  # Usage
  model = ScanMLP(3, n_layers=4, rngs=nnx.Rngs(0))
  output = model(jnp.array([1.0, 2.0, 3.0]))

  Summary

  - nnx.Carry: Thread through all iterations
  - Integer (0,1,2...): Which axis to iterate/stack over
  - None: Broadcast same value to all iterations
  - Main advantage: Can scan over Module objects, not just arrays

  Your understanding is perfect! The key insight is nnx.scan makes it elegant to apply sequences of neural network layers.

**Comment by mohsinm-dev** [comment]
@ak24watch 

Your understanding is **100% correct**! See the examples.

  ### Key Differences
  - **`jax.lax.scan`**: Works with arrays only, manual state management
  - **`nnx.scan`**: Works with NNX Modules, automatic state management

### Simple Example

```python
  import jax.numpy as jnp
  from flax import nnx

  # Running sum example
  @nnx.scan(in_axes=(nnx.Carry, 0), out_axes=(nnx.Carry, 0))
  def running_sum(carry, x):
      new_carry = carry + x
      output = new_carry
      return new_carry, output

  numbers = jnp.array([1.0, 2.0, 3.0, 4.0])
  final_sum, all_sums = running_sum(0.0, numbers)

  print(final_sum)    # 10.0
  print(all_sums)     # [1.0, 3.0, 6.0, 10.0]

  class Block(nnx.Module):
      def __init__(self, dim, rngs):
          self.linear = nnx.Linear(dim, dim, rngs=rngs)

      def __call__(self, x):
          return nnx.relu(self.linear(x))

  class ScanMLP(nnx.Module):
      def __init__(self, dim, n_layers, rngs):
          # Create vectorized layers
          @nnx.split_rngs(splits=n_layers)
          @nnx.vmap(axis_size=n_layers)
          def create_block(rngs):
              return Block(dim, rngs=rngs)

          self.layers = create_block(rngs)

      def __call__(self, x):
          # Scan through layers
          @nnx.scan(in_axes=(nnx.Carry, 0), out_axes=(nnx.Carry, 0))
          def apply_layer(x, layer):
              result = layer(x)
              return result, result

          final_x, intermediates = apply_layer(x, self.layers)
          return final_x
  
  # Usage
  model = ScanMLP(3, n_layers=4, rngs=nnx.Rngs(0))
  output = model(jnp.array([1.0, 2.0, 3.0]))
```
  
Summary
  - nnx.Carry: Thread through all iterations
  - Integer (0,1,2...): Which axis to iterate/stack over
  - None: Broadcast same value to all iterations
  - Main advantage: Can scan over Module objects, not just arrays

**Comment by vfdev-5** [comment]
@mohsinm-dev thanks for helping to answer this question! (feels like an llm answer :)  anyway)

@ak24watch By the way, I'm updating the docstring for nnx.scan here: https://github.com/google/flax/pull/4930 
Hopefully, it will be much better now.
